### PR TITLE
Slim 1095 replace soapui tests for bundles

### PR DESCRIPTION
**Pull request by smvdheijden**

### CODE REVIEW

**Comment by smvdheijden** [comment]
Retest this please

**Comment by bvdzwet** [comment]
As this is the last part of replacing the SOAP UI way of Cucumber testing with Java SOAP clients, I would think SOAP UI related dependencies would be removed as well.

The parent-integration-tests/pom.xml and cucumber-tests-platform-smartmetering/pom.xml have some dependency, plugin and repository configuration pointing to SOAP UI related artifacts that should have become obsolete I think.

The Cucumber tests platform smartmetering still have a WsdlProjectFactory for use with SOAP UI.

Some of the classes use StringUtils from Apache commons-lang, which comes in as a transitive dependency through SOAP UI related artifacts. It appears this could be replaced by commons-lang3 (adding the dependency, for which the version is already in the parent POM, and updating the import statements).

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    private AlarmNotification getDefaultAlarmNotification() {
+        final AlarmNotification alarmNotification = new AlarmNotification();
+        alarmNotification.setAlarmType(DEFAULT_ALARM_TYPE);
+        alarmNotification.setEnabled(true);
+        return alarmNotification;
+    }
+
+    private AlarmNotification getAlarmNotification(final Map<String, String> parameters, final int index) {
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java]: Consider using the indexed methods from SettingsHelper to ensure consistent handling of indexed values in Cucumber data tables.

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        final AlarmNotification alarmNotification = new AlarmNotification();
+        alarmNotification.setAlarmType(this.getAlarmType(parameters, index));
+        alarmNotification.setEnabled(this.getEnabled(parameters, index));
+        return alarmNotification;
+    }
+
+    private AlarmType getAlarmType(final Map<String, String> parameters, final int index) {
+        final String key = PlatformSmartmeteringKeys.ALARM_TYPE + "_" + index;
+        if (parameters.containsKey(key)) {
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java]: Consider using the methods with a default value parameter from Helpers to avoid repeating the if-contains-key-else-return-default code.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ClockConfiguration.feature [hunk 3]
```diff
@@ -1,3 +1,4 @@
+# @TestThis
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ClockConfiguration.feature]: The `@TestThis` tag does not belong in the pull request I think.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature [hunk 4]
```diff
@@ -10,51 +10,38 @@ Feature: SmartMetering Bundle
...
+    And the bundle request contains a set activity calendar action
+    And the bundle request contains a get configuration object action
+    And the bundle request contains a set alarm notifications action
+    And the bundle request contains a set configuration object action
+    And the bundle request contains a set push setup alarm action
+    And the bundle request contains a synchronize time action
+    When the bundle request is received
+    Then the number of responses in the bundle response should match the number of actions in the bundle request
+    And the order of the responses in the bundle response should match the order of actions in the bundle request
+    #    And the bundle response should contain a find events response
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature]: It looks like these commented-out checks are handled by "And the order of the responses in the bundle response should match the order of actions in the bundle request".
If there is something left TODO it should be made clear or done, and these commented-out lines shouldn't be here I think.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature [hunk 5]
```diff
@@ -10,51 +10,38 @@ Feature: SmartMetering Bundle
...
+    #    And the bundle response should contain a get administrative status response
+    #    And the bundle response should contain a get periodic meter reads response
+    #    And the bundle response should contain a set activity calendar response
+    #    And the bundle response should contain a get configuration object response
+    #    And the bundle response should contain a set alarm notifications response
+    #    And the bundle response should contain a set configuration object response
+    #    And the bundle response should contain a set push setup alarm response
+    #    And the bundle response should contain a synchronize time response
+    
+
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature]: There's some trailing whitespace on these last two lines that should be avoided.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledFindEvents.feature [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a find events action with parameters
+      | EventLogCategory | FRAUD_DETECTION_LOG      |
+      | From             | 2014-01-01 00:00:00 |
+      | Until            | 2017-01-01 00:00:00 |
+    When the bundle request is received
+    Then the bundle response should contain a find events response with values
+      | Result | OK |
+
+
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledFindEvents.feature]: There's some trailing whitespace on the last line that should be avoided.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+      | DeviceType           | SMART_METER_E     |
+
+  Scenario: Get actual meter reads of a device in a bundle request
+    Given a bundle request
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a get actual meter reads action
+    When the bundle request is received
+    Then the bundle response should contain a get actual meters reads response with values
+      | Result | OK |
+
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature]: There's some trailing whitespace on the last line that should be avoided. (Check other features as well, as I won't be repeating this for all the added "Bundled"-features.)

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    Given a dlms device
+      | DeviceIdentification | TEST1024000000001 |
+      | DeviceType           | SMART_METER_E     |
+
+  Scenario: Get actual meter reads of a device in a bundle request
+    Given a bundle request
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a get actual meter reads action
+    When the bundle request is received
+    Then the bundle response should contain a get actual meters reads response with values
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature]: There is no step implementation code for this 'Then'-step. Shouldn't this instead be: "the bundle response should contain a get actual meter reads response with values" (meter reads instead of meters reads)?

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetAdministrativeStatus.feature [hunk 9]
```diff
@@ -0,0 +1,18 @@
+@SmartMetering @Platform
+Feature: SmartMetering Bundle - GetAdministrativeStatus
+  As a grid operator 
+  I want to be able to get administrative status from a meter via a bundle request
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetAdministrativeStatus.feature]: The phrase "via a bundle request" appears here, but is not consistently used throughout the "Bundled"-features, it might be cleaner to do so.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetPeriodicMeterReads.feature [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+  Scenario: Get periodic meter reads of a device in a bundle request
+    Given a bundle request
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a get periodic meter reads action with parameters
+      | PeriodType | DAILY      |
+      | BeginDate  | 2016-01-01 |
+      | EndDate    | 2017-01-01 |
+    When the bundle request is received
+    Then the bundle response should contain a get periodic meters reads response with values
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetPeriodicMeterReads.feature]: This then-step does not have a matching step implementation, should it instead be: "Then the bundle response should contain a get periodic meter reads response with values" (meter reads instead of meters reads)?

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetAlarmNotifications.feature [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+      | DeviceIdentification | TEST1024000000001 |
+      | DeviceType           | SMART_METER_E     |
+
+  Scenario: Set alarm notifications on a device in a bundle request
+    Given a bundle request
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a set alarm notifications action with parameters
+      | AlarmNotificationCount |             2 |
+      | AlarmType_1            | POWER_UP      |
+      | AlarmTypeEnabled_1     |             true |
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetAlarmNotifications.feature]: You could format this file to get the columns a bit neater aligned in this step.

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetSpecialDays.feature [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  Scenario: Set special days on a device in a bundle request
+    Given a bundle request
+      | DeviceIdentification | TEST1024000000001 |
+    And the bundle request contains a set special days action with parameters
+      | SpecialDayCount  |          1 |
+      | SpecialDayId_1   |          3 |
+      | SpecialDayDate_1 | FFFFFFFFFF |
+    When the bundle request is received
+    Then the bundle response should contain a set special days response with values
+      | Result               | OK                |
```
Comment (bvdzwet) [inline cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetSpecialDays.feature]: Was something in this last step present before and removed, causing the amount of whitespace after Result and OK that is quite a bit larger than usual?

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/config/SmartMeteringMapper.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
+package com.alliander.osgp.cucumber.platform.smartmetering.config;
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/config/SmartMeteringMapper.java]: License header added

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledFindEventsSteps.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        final BundleRequest request = (BundleRequest) ScenarioContext.current()
+                .get(PlatformSmartmeteringKeys.BUNDLE_REQUEST);
+
+        final FindEventsRequest action = new FindEventsRequestBuilder().withDefaults().build();
+
+        this.addActionToBundleRequest(request, action);
+    }
+
+    @Given("^the bundle request contains a find events action with parameters$")
+    public void theBundleRequestContainsAFindEventsAction(final Map<String, String> parameters) throws Throwable {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledFindEventsSteps.java]: The with parameters step method allows for more flexibility, it can be used for non-happy scenario testing as well...

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledFindEventsSteps.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    @Then("^the bundle response should contain a find events response$")
+    public void theBundleResponseShouldContainAFindEventsResponse() throws Throwable {
+
+        final Response response = this.getNextBundleResponse();
+
+        assertTrue("Not a valid response", response instanceof FindEventsResponse);
+    }
+
+    @Then("^the bundle response should contain a find events response with values$")
+    public void theBundleResponseShouldContainAFindEventsResponse(final Map<String, String> values) throws Throwable {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledFindEventsSteps.java]: I would have expected a result OK in the find events response, but it looks like it's always null...
When no events are found, all fields in the response are null?

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetActualMeterReadsSteps.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Then("^the bundle response should contain a get actual meter reads response$")
+    public void theBundleResponseShouldContainAGetActualMeterReadsResponse() throws Throwable {
+
+        final Response response = this.getNextBundleResponse();
+
+        assertTrue("Not a valid response", response instanceof ActualMeterReadsResponse);
+    }
+
+    @Then("^the bundle response should contain a get actual meter reads response with values$")
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetActualMeterReadsSteps.java]: It is used in the BundledGetActualMeterReads,feature file

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    public void theBundleRequestContainsAGetPeriodicMeterReadsActionWithParameters(final Map<String, String> parameters)
+            throws Throwable {
+
+        final BundleRequest request = (BundleRequest) ScenarioContext.current()
+                .get(PlatformSmartmeteringKeys.BUNDLE_REQUEST);
+
+        final GetPeriodicMeterReadsRequest action = new GetPeriodicMeterReadsRequestBuilder()
+                .fromParameterMap(parameters).build();
+
+        this.addActionToBundleRequest(request, action);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetPeriodicMeterReadsSteps.java]: Moved retrieving the `BundleRequest` from the scenario context to the `addActionToBundleRequest` method, and dropped the parameter.

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledSetActivityCalendarSteps.java [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+                .get(PlatformSmartmeteringKeys.ACTIVITY_CALENDAR);
+
+        final int dayId = Integer.parseInt(parameters.get(PlatformSmartmeteringKeys.DAY_PROFILE_DAY_ID));
+        final int actionCount = Integer.parseInt(parameters.get(PlatformSmartmeteringKeys.DAY_PROFILE_ACTION_COUNT));
+
+        final DayProfile profile = new DayProfile(dayId);
+
+        for (int i = 1; i <= actionCount; i++) {
+
+            final String start = parameters.get(PlatformSmartmeteringKeys.DAY_PROFILE_START_TIME + "_" + i);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledSetActivityCalendarSteps.java]: Updated code

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+import com.alliander.osgp.adapter.ws.schema.smartmetering.management.EventLogCategory;
+import com.alliander.osgp.cucumber.platform.helpers.DateConverter;
+import com.alliander.osgp.cucumber.platform.smartmetering.PlatformSmartmeteringKeys;
+
+public class FindEventsRequestBuilder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FindEventsRequestBuilder.class);
+
+    private static final String DATE_FORMAT = "yyyy-MM-dd hh:mm:ss";
+    private static final SimpleDateFormat SDF = new SimpleDateFormat(DATE_FORMAT);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java]: Renamed to SIMPLE_DATE_FORMAT and DATE_PATTERN

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+
+    private static final EventLogCategory DEFAULT_EVENT_LOG_CATEGORY = EventLogCategory.FRAUD_DETECTION_LOG;
+    private static final DateTime DEFAULT_FROM = new DateTime(2016, 1, 1, 0, 0, 0, DateTimeZone.UTC);
+    private static final DateTime DEFAULT_UNTIL = new DateTime(2017, 1, 1, 0, 0, 0, DateTimeZone.UTC);
+
+    private EventLogCategory eventLogCategory;
+    private XMLGregorianCalendar from;
+    private XMLGregorianCalendar until;
+
+    public FindEventsRequestBuilder withDefaults() {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java]: I've added the withDefaults method, as I thought it would be more explicit...

[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java]: Updated code in this class as well as in some other builder classes

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetActivityCalendarRequestBuilder.java [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+import com.alliander.osgp.cucumber.platform.smartmetering.support.ws.smartmetering.bundle.activitycalendar.DayProfileAction;
+import com.alliander.osgp.cucumber.platform.smartmetering.support.ws.smartmetering.bundle.activitycalendar.SeasonProfile;
+import com.alliander.osgp.cucumber.platform.smartmetering.support.ws.smartmetering.bundle.activitycalendar.WeekProfile;
+
+public class SetActivityCalendarRequestBuilder {
+    private static final ActivityCalendar DEFAULT_ACTIVITY_CALENDAR = new ActivityCalendar("ACTCAL01",
+            "FFFFFFFEFFFFFFFFFF000000");
+
+    static {
+        final SeasonProfile sseasonProfile = new SeasonProfile("SEASON01", "FFFF0101FFFFFFFFFF000000", "WEEK0001");
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetActivityCalendarRequestBuilder.java]: done

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetActivityCalendarRequestBuilder.java [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        final WeekProfile weekProfile = new WeekProfile("WEEK0001", 0, 0, 0, 0, 0, 0, 0);
+        final DayProfileAction dayprofileAction = new DayProfileAction("00000000", 1);
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetActivityCalendarRequestBuilder.java]: done

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java (hunk 1 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java]: updated code

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java (hunk 2 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/SetAlarmNotificationsRequestBuilder.java]: Updated code

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ClockConfiguration.feature (hunk 3 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/ClockConfiguration.feature]: Removed

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature (hunk 4 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature]: Initially I had implemented the individual response checks.
After adding the number of and order of steps these were no longer needed.
Removed the commented-out lines

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature (hunk 5 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/SmartMeteringBundle.feature]: Removed

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledFindEvents.feature (hunk 6 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledFindEvents.feature]: Removed

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature (hunk 8 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature]: Fixed typo

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature (hunk 7 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetActualMeterReads.feature]: Removed

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetPeriodicMeterReads.feature (hunk 10 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetPeriodicMeterReads.feature]: Fixed typo

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetAlarmNotifications.feature (hunk 11 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetAlarmNotifications.feature]: Formatted

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetSpecialDays.feature (hunk 12 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledSetSpecialDays.feature]: Formatted

File: cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetAdministrativeStatus.feature (hunk 9 above)
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/bundle/BundledGetAdministrativeStatus.feature]: Added  "via a bundle request" where needed

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+
+    public FindEventsRequest build() {
+        final FindEventsRequest request = new FindEventsRequest();
+        request.setEventLogCategory(this.eventLogCategory);
+        request.setFrom(this.from);
+        request.setUntil(this.until);
+        return request;
+    }
+
+    private EventLogCategory getEventLogCategory(final Map<String, String> parameters) {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java]: Refactored to a one-liner using `Helpers.getEnum`

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/GetPeriodicMeterReadsRequestBuilder.java [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+            return PeriodType.fromValue(parameters.get(PlatformSmartmeteringKeys.KEY_PERIOD_TYPE));
+        }
+        LOGGER.debug("Key for period type not found in parameters, using default value.");
+        return DEFAULT_PERIOD_TYPE;
+    }
+
+    private XMLGregorianCalendar getBeginDate(final Map<String, String> parameters) {
+        if (parameters.containsKey(PlatformSmartmeteringKeys.KEY_BEGIN_DATE)) {
+            try {
+                final Date beginDate = SDF.parse(parameters.get(PlatformSmartmeteringKeys.KEY_BEGIN_DATE));
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/GetPeriodicMeterReadsRequestBuilder.java]: Refactored to use the Helpers.getDate method

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+import org.joda.time.DateTimeZone;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.alliander.osgp.adapter.ws.schema.smartmetering.bundle.FindEventsRequest;
+import com.alliander.osgp.adapter.ws.schema.smartmetering.management.EventLogCategory;
+import com.alliander.osgp.cucumber.platform.helpers.DateConverter;
+import com.alliander.osgp.cucumber.platform.smartmetering.PlatformSmartmeteringKeys;
+
+public class FindEventsRequestBuilder {
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/support/ws/smartmetering/bundle/FindEventsRequestBuilder.java]: Perhaps we could use a generic interface like below:
`public interface RequestFactory<T> {
    T createWithDefaults();
    T createFromParameterMap(Map<String, String> parameters);
}`

For now, I'll  leave it as is, as it might take too much time to refactor...

File: cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetAdministrativeStatusSteps.java [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Then("^the bundle response should contain a get administrative status response$")
+    public void theBundleResponseShouldContainAGetAdministrativeStatusResponse() throws Throwable {
+
+        final Response response = this.getNextBundleResponse();
+
+        assertTrue("Not a valid response", response instanceof AdministrativeStatusResponse);
+    }
+
+    @Then("^the bundle response should contain a get administrative status response with values$")
```
[the remark this answers is not in the document]
Comment (smvdheijden) [reply cucumber-tests-platform-smartmetering/src/test/java/com/alliander/osgp/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetAdministrativeStatusSteps.java]: Apparently most operations retrieving data don't use the result field. So it is not possible to test on this result.
Personally, I would prefer to use the result field for all operations, whether retrieving or setting data... 
For now I've removed most unused with values step methods...
